Add aten mkldnn conv2d operator

bddppq opened the pull request:
Differential Revision: D14857513

gchanan commented on test/test_mkldnn.py:
```diff
@@ -88,5 +89,23 @@ def test_repr(self):
+                                 kernel_size=3, stride=2, padding=1).float()
+
+        mkldnn_conv2d = copy.deepcopy(conv2d)
+        mkldnn_conv2d.weight.data = mkldnn_conv2d.weight.data.to_mkldnn()
```
can you do:
```
with torch.no_grad():
   mkldnn_conv2d.weight = mkldnn_conv2d.weight.to_mkldnn()
   mkldnn_conv2d.bias = mkldnn_conv2d
```

gchanan reviewed (commented):
is the plan to use these generally with modules?  Because it's kind of clunky as is, but not clunky with functions.

dzhulgakov commented:
I guess we could introduce Module.to_mkldnn or a standalone function for it that handles common modules (potentially by putting some method on module implementations themselves)

bddppq commented:
@pytorchbot retest this please

jgong5 commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -35,43 +35,78 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
+      dummy_scale,
+      ideep::descriptor_group::attr_t{},
+      ideep::algorithm::convolution_direct,
+      ideep::prop_kind::forward_inference);
```
Suggest to use `ideep::prop_kind::forward` instead since the same function would be used in both training and inference.

bddppq replied:
Good point. will change it.
Actually I'm curious what's the perf difference between forward_inference and forward_training for ideep conv?

jgong5 replied:
@bddppq So far, the only difference lies in Winograd conv. For inference, weight transformation is only done once and transformed weight can be cached in MKL-DNN opaque format (the same way as other blocking formats). For training, weight is transformed all the time.

jgong5 commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -35,43 +35,78 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
+      dummy_scale,
+      ideep::descriptor_group::attr_t{},
+      ideep::algorithm::convolution_direct,
+      ideep::prop_kind::forward_inference);
```
ditto

ezyang commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -301,6 +301,10 @@ at::Tensor conv1d(
+  if (input.type_id() == MkldnnCPUTensorId()) {
+    return at::mkldnn_convolution(
+        input, weight, bias, padding, stride, dilation, groups);
+  }
```
It's kind of weird this only applies to conv2d.

gchanan replied:
to pile on to @ezyang's point here -- this is also skipping all the useful error checking and utility code that already exists for convolution.  Consider the difference in the experience here:
```
>>> F.conv2d(torch.randn(1,8,4,3,3), torch.randn(1,4,5,5), padding=1)
RuntimeError: Expected 4-dimensional input for 4-dimensional weight [1, 4, 5, 5], but got 5-dimensional input of size [1, 8, 4, 3, 3] instead
```

```
>>> F.conv2d(torch.randn(1,8,4,3,3).to_mkldnn(), torch.randn(1,4,5,5).to_mkldnn(), padding=1)
RuntimeError: std::exception
```

there's already logic in here for when to use MKLDNN -- why can't we just modify that to take into account if we have an MKLDNNTensor?

Also, I believe if you do it this way you shouldn't have to reproduce `expand_param_if_needed` and possibly not even `conv_output_size`.

bddppq replied:
1. mkldnn is not supposed to handle all convNd, that's why here I only added it for conv2d
2. there are a lot of unnecessary things done before going into the use_mkldnn branch and I want to avoid that for mkldnn (e.g. input.contiguous). To be fair the dispatch should really be put in the native_functions.yaml and let codegen handles it, but because of varies kind of autograd reason we now need to put it into the op itself, so it's better to put this dispatch at the very beginning to make it clear (instead of hiding this dispatch somewhere in the already complicated _convolution).
3. you still need conv_output_size, because you need to pass it to ideep

bddppq replied:
Oh also we intentionally to call at::mkldnn_convolution here (instead of directly calling to the ideep_conv2d c++ function) because we need to use the codegen VariableType::mkldnn_convolution to do unwrap.

gchanan replied:
> Oh also we intentionally call at::mkldnn_convolution here (instead of directly calling to the ideep_conv2d c++ function) because we need to use the codegen VariableType::mkldnn_convolution to do unwrap.

Calling `mkldnn_convolution` in this file is correct for the reason that you note.  The question is about when you call it.

As to your points:
> mkldnn is not supposed to handle all convNd, that's why here I only added it for conv2d

Sure, that's fine, you don't need to call it in all cases.  The point is to just move the dispatch logic to the place that already exists so you get error checking and such for free.  None of the existing code unconditionally calls other kernel -- it checks the parameters and dispatches to the 'best' kernel or errors out if it can't be supported.

> there are a lot of unnecessary things done before going into the use_mkldnn branch and I want to avoid that for mkldnn (e.g. input.contiguous).

It sounds like we should fix those because they are wrong?

> To be fair the dispatch should really be put in the native_functions.yaml and let codegen handles it, but because of varies kind of autograd reason we now need to put it into the op itself, so it's better to put this dispatch at the very beginning to make it clear (instead of hiding this dispatch somewhere in the already complicated _convolution).

This is just moving the problem.  You would still need to reproduce all the error checking / setup code.

ezyang commented:
I'm LGTM'ing and resigning from the native part of the review. If you need me to review the binding code as well please say so.

gchanan commented on test/test_mkldnn.py:
```diff
@@ -75,18 +76,38 @@ def test_detach(self):
 
         detach = root.detach()
-        self.assertEquals((4, 5), detach.size())
+        self.assertEqual((4, 5), detach.size())
```
what's going on here?  Are both Equal and Equals usable?

bddppq replied:
Yes. I change it because I got a warning saying Equals is being deprecated and here should change to Equal

gchanan commented on test/test_mkldnn.py:
```diff
@@ -75,18 +76,38 @@ def test_detach(self):
+                                     bias=bias).float()
+            mkldnn_conv2d = copy.deepcopy(conv2d).to_mkldnn()
+
+            with torch.no_grad():
```
why do you need the no_grad?

bddppq replied:
Lol didn't you ask me to use to_grad in an comment above? https://github.com/pytorch/pytorch/pull/19075#discussion_r273675706

gchanan replied:
Sorry for not being clear.

The recommendation was to use `no_grad` instead of `.data` -- that's just a general recommendation.

As for `.data` / `no_grad` itself, I'm trying to understand if you actually need it.  Do you?

bddppq replied:
No it is not needed. Plus I now have added a to_mkldnn function at Module and used `_apply` to call to_mkldnn which handles the grad/.data thing well. I will remove the no_grad here.

gchanan commented on aten/src/ATen/native/mkldnn/Utils.h:
```diff
@@ -0,0 +1,25 @@
+
+#include <c10/util/ArrayRef.h>
+#include <vector>
+#include <ideep.hpp>
```
this doesn't need to be here, right?  There's nothing ideep specific here.

bddppq replied:
yes will remove

gchanan commented on aten/src/ATen/native/mkldnn/Utils.cpp:
```diff
@@ -0,0 +1,36 @@
+  }
+}
+
+std::vector<int64_t> conv_output_size(
```
why did this move?  This has nothing to do with mkldnn, right?

bddppq replied:
this was in mkldnn/Conv.cpp before and only used in mkldnn_convolution

bddppq commented:
@pytorchbot retest this please

mingfeima commented:
Hi @bddppq i am working with @Jianhui-Li on #16038.
Here are a few inputs from our side, just for your reference:

1. Probably leave `at::mkldnn_convolution()` with the rest of acceleration libraries (`cudnn`, `miopen`) at `_convolution()` is a better idea.
2. Maybe need to support some factory tensor operations such as `is_contiguous()`, `contiguous()`, `clone` etc. on `MkldnnCPUTensorId` so as to pass some check, for example, the contiguous check.
and we are going to need `view` in case we want to use `mkldnn` for `nn.Linear()`
taking the most simple example of [alexnet](https://github.com/pytorch/vision/blob/master/torchvision/models/alexnet.py#L46). We need `view` to support `MkldnnCPUTensorId` otherwise it's going to be runtime error.
Or we add `to_dense` before `view` and `to_dense` after, but in that case, `mkldnn` blocked format is not passed on and the performance may be sub optimal.
3. Few humble thoughts on weight cache: my old idea is to do inplace reorder on the fly, just like [caffe2](https://github.com/pytorch/pytorch/blob/master/caffe2/ideep/operators/conv_op.cc#L63) did. @Jianhui-Li mentioned you plan to do this when calling `model.to_mkldnn()` which is probably a better idea, the semantics is more clear. Yet this might involves some dependencies: a) mkldnn library needs to provide helper_function to return desired memory descriptor out of convolution; b) need to expose mkldnn memory descriptor in python perheps, e.g. `tensor.to_mkldnn(blocked_desc)`, but now the struct of memory descriptor can't exposed in `native_functions.yaml`.
4. In case we need to cover `training`, we may need optimizer to understand `MkldnnCPUTensorId`. We have two options here: 
a) fill up every math op with `mkldnn` tensor. This solution is general but requires lots of efforts. Also math op can be used in vast scenarios which increase the effort of maintaining.
b) also we can use a shortcut, e.g. `mkldnn_sgd_step()` and `mkldnn_adam_step`, this is similar to current `sparse` optimizer design. And it is performance beneficial since it is fused. the drawback is that it doesn't support customized optimizer.
5. Some trivial parts: possibly we can move headers from `ATen/native/mkldnn` to `ATen/mkldnn` along with the util functions that are not meant to be exposed in `native_functions.yaml`. Similar to other acceleration library folders.

We have done some similar job on our side and i post the code at this branch [mkldnn/tensor](https://github.com/mingfeima/pytorch/tree/mkldnn/tensor), both forward and backward is updated. this branch holds `nn.Conv2d` and `nn.ConvTransposed2d`. Feel free to take any fragments.

Per your request, we will provide `batchnorm` and `addmm` **ASAP**. We have done `batchnorm`, `pooling`, `relu` etc in `ideep`, just needs some rebase, testing and sum up things together.

About `addmm`, probably we need to hook under `nn.Linear` instead of `addmm`? `addmm` is very widely used, for example even `bmm()` calls `addmm`. The semantics is much broader than what `mkldnn` provides...

mingfeima commented:
Some discussion on `mkldnn` in `nn.Linear()`:
1. **easiest way**: call `to_dense` before `view`, a.k.a use cpu tensor for `nn.Linear()`, sgemm will be carried out by `mkl`. And do noting else. The performance would be the same as using `mkldnn` for `nn.Linear()` for float32 (`mkldnn` will be faster for int8)
2. **another way**: add `torch.linear` in `functional.py`. We will propose a small PR to further illustrate this. Reason is `addmm` is very widely used, for example even `bmm` calls `addmm`. The semantics is much broader than what mkldnn provides...
```bash
before: nn.Linear() -> torch.addmm() -> mkl
after: nn.Linear() -> torch.linear() -> mkldnn     # only if the tensor is mkldnn; use addmm for other tensortypes
```

Any feedback is welcome.

apaszke commented on torch/nn/modules/module.py:
```diff
@@ -273,6 +273,9 @@ def cpu(self):
         """
         return self._apply(lambda t: t.cpu())
 
+    def to_mkldnn(self):
```
No, we can't do that!!!! We need a unified interface for all backends!! This is the main subclass for half of our library, don't put random functions in there!

apaszke reviewed (changes requested):
`to_mkldnn` needs to go! We need to have a solution that will work for other backends too, and we already have it -- use the `.to()` function.

EDIT: Giving this more thought, if you want to put the tensors into MKLDNN layout, we actually shouldn't use `.to()` because that's for devices, not layouts. We should have a different function that lets us select the layout, but it should support open registration and all similarly to `to()`. Otherwise it will become a mess. We don't have `to_cudnn()` even though there are some modules which require us to reformat our weights.

soumith commented:
instead of `to_mkldnn`, we need to overload `to(layout='mkldnn')`, rather than adding a method to the frontend. It should get us to the same end-result.

gchanan commented:
@mingfeima can you create a new issue with the discussion about `view`, `nn.Linear`, etc.  There's enough going on in this issue that overloading it with future plans is not a good idea.

gchanan commented:
> EDIT: Giving this more thought, if you want to put the tensors into MKLDNN layout, we actually shouldn't use .to() because that's for devices, not layouts. We should have a different function that lets us select the layout, but it should support open registration and all similarly to to(). Otherwise it will become a mess. We don't have to_cudnn() even though there are some modules which require us to reformat our weights.

> instead of to_mkldnn, we need to overload to(layout='mkldnn'), rather than adding a method to the frontend. It should get us to the same end-result.


We _kind_ of currently support layout in `to(...)` in C++, just because TensorOptions takes a layout (why?), although we currently check that you aren't trying to convert and we don't bind it to python (phew).

One complication is that in general layouts may take extra parameters.  For example `to_sparse` takes the number of sparse dimensions.  You probably wouldn't want to take those as kwargs, because that gets a little crazy (what if a parameter of `to` overlaps with a parameter of the conversion function?).

I think the two most obvious possibilities here are:
1) we have a generic layout-only conversion function that that takes kwargs and passes them on to a per-layout-type conversion function (e.g. `to_sparse`) that can do its own parsing of the kwargs.
2) we support "layout descriptors" that are capable of completely specifying layout conversions.  For example, a sparse layout with 2 sparse dimensions could be something like "sparse_coo/2".  The user could call a per-layout-type descriptor factory to create these.

gchanan commented:
btw for unblocking the PR, you can just call `apply` or just test the function instead of adding `to_mkldnn` to the Module.

mingfeima commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -35,43 +35,78 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
+  if (input.type_id() == MkldnnCPUTensorId()) {
+    return _ideep_conv2d(input, weight, bias, padding, stride, dilation, groups);
+  }
+
```
1. the logic is for `MkldnnCPUTensorId()` go `_ideep_conv2d` and else (`CPUTensorId`) go `mkldnn_convolution`, actually it is not necessary and they can be merged into one function.
2. `ideep` is not a standalone library, it is a header that wraps up `mkldnn`. Their function are identical so there is no need to keep two sets of implementations. The current code may mislead people that `ideep` and `mkldnn` are two parallel libraries.
3. The existing code `mkldnn_convolution`, `mkldnn_convolution_backward_xxx` calls `mkldnn` directly. Since we need to switch them to `ideep`. I suggest switch them all and remove `Runtime.h` as well as other direct `mkldnn` callings. You may take #18460 as a reference.

mingfeima commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -35,43 +35,78 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
 
-// Often written as 2 + max_dim (extra dims for batch size and channels)
-constexpr int max_dim = 3;
+  auto stride_vec = expand_param_if_needed(stride, 2);
```
`expand_param_if_needed` is not needed, in case we still call `mkldnn_convolution` from `_convolution()`:
1. inside `use_mkldnn` from `Convolution.cpp`, it will regulate mkldnn for 2d convolution only, and 3d convolution will go `THNN` for now. and 1d convolution reuses 2d logic. So `stride, padding, dilation` can be only of vector of 2 here.
2. in near future, when `ideep` can handle 3d convolution, `mkldnn_convolution` will be able to handle both 2d and 3d in one place, just like `cudnn_convolution` did.

gchanan commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -35,43 +35,78 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
 #include <ATen/mkldnn/Runtime.h>
+#include <ATen/native/mkldnn/MKLDNNCommon.h>
+#include <ATen/native/mkldnn/Utils.h>
 
```
Why do we need to do 2?

bddppq committed:
Add aten mkldnn conv2d operator

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/19075

Differential Revision: D14857513

fbshipit-source-id: 12254d7c13ccc821cb558de8c6eeb8443035dd0c

bddppq commented:
Sorry folks in order to export my whole stack of diffs I need to create a new pull request for this one: https://github.com/pytorch/pytorch/pull/19204. I'm closing this one now and will address review comments in the new pull request.
